### PR TITLE
Proxy blueprint examples

### PR DESCRIPTION
**Pull request by lrubasze**
## Summary

This folder includes Proxy blueprint examples for `Oracle` component (available in `oracle_v1`, `oracle_v2`
and `oracle_v3` blueprints).
It is assumed that `Oracle` component includes:
- public methods, eg. `get_price()`
- protected methods, eg. `set_price()`
Below examples take above into account.

Proxy blueprint examples:
- `oracle_proxy_with_global`
    - It assumes that signatures of `Oracle` component proxied methods will remain unchanged
    (thus not compliant with `oracle_v3`)
    - This proxy works with `Oracle` instantiated as a global component
    - It can call only public methods from `Oracle` component

- `oracle_proxy_with_owned`
    - It assumes that signatures of `Oracle` component proxied methods will remain unchanged
    (thus not compliant with `oracle_v3`)
    - It works with `Oracle`  instantiated as an owned component (it must be instantiated by this proxy)
    - It can call any method from owned `Oracle` component

- `oracle_generic_proxy_with_global`
    - It is a generic proxy which can call any method with any arguments from configured component
    - It works with component instantiated as a global component
    - It can call only public methods from configured component

NOTE!
There is no `oracle_generic_proxy_with_owned` example because proxy generic `call_method()` shall be public.
And since the proxy owns the component it can call any method of the owned component and this is not acceptable,
since some methods eg. `set_price()` shall be protected.

### CODE REVIEW

File: radix-engine-tests/assets/blueprints/oracle_proxy_basic/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    struct OracleProxy {
+        // Either of below fields must be set in order to use Proxy.
+        // - Use `oracle_global_address` if Oracle is a global component.
+        // - Use `oracle_owned_address` if Oracle is an owned component.
+        oracle_global_address: Option<Global<AnyComponent>>,
+        // Vector used here, because it is not possible to drop Owned value.
+        // Only the last added address will be used, previous addresses will be kept, despite not being used.
+        oracle_owned_address: Vec<Owned<AnyComponent>>,
```
[the remark this answers is not in the document]
Comment (iamyulong) [reply radix-engine-tests/assets/blueprints/oracle_proxy_basic/src/lib.rs:25]: Also, may be we can move generic proxy here as well? Like`GenericProxyWithGlobal` and `GenericProxyWithOwned`.

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    struct OracleProxy {
+        oracle_global_address: Option<Global<AnyComponent>>,
+    }
+
+    // This example demostrate a proxy working with:
+    // - Oracle as a global component
+    //   Proxy can call only public methods of Oracle component
+    impl OracleProxy {
+        pub fn instantiate_global(
```
Comment (iamyulong) [inline radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs:25]: Minor: rename `instantiate_and_globalize`

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+                    .handle()
+                    .as_node_id(),
+                "get_price",
+                scrypto_args!(base, quote),
+            );
+            scrypto_decode(&result).unwrap()
+        }
+
+        // This method will fail if Oracle is a global component
+        pub fn set_price(&self, base: ResourceAddress, quote: ResourceAddress, price: Decimal) {
```
Comment (iamyulong) [inline radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs:75]: Minor: should we just remove this method?

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+                    .expect("Oracle not initialized")
+                    .handle()
+                    .as_node_id(),
+                "get_price",
+                scrypto_args!(base, quote),
+            );
+            scrypto_decode(&result).unwrap()
+        }
+
+        // This method will fail if Oracle is a global component
```
Comment (iamyulong) [inline radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs:85]: Remove this comment.

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        // Vector used here, because it is not possible to drop Owned value.
+        // Only the last added address will be used, previous addresses will be kept, despite not being used.
+        oracle_owned_address: Vec<Owned<AnyComponent>>,
+    }
+
+    // This example demostrate a proxy working with:
+    // - Oracle as an owned component (instantiated by proxy)
+    //   Proxy can call any method from owned Oracle component
+    impl OracleProxy {
+        pub fn instantiate_global(
```
Comment (iamyulong) [inline radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs:27]: Minor: `instantiate_and_globalize`

**Review by iamyulong — APPROVED** [review]
LGTM

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs (hunk 2 above)
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_global/src/lib.rs]: Done

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs (hunk 4 above)
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs]: Done

File: radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs (hunk 5 above)
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/oracle_proxies/oracle_proxy_with_owned/src/lib.rs]: Done

File: radix-engine-tests/assets/blueprints/oracle_proxy_basic/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/oracle_proxy_basic/src/lib.rs]: `oracle_proxy_basic` splitted into `oracle_proxy_with_global` and `oracle_proxy_with_owned`.
`generic_proxy` renamed to `oracle_generic_proxy_with_owned`

[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/oracle_proxy_basic/src/lib.rs]: As agreed generic proxy with Owned does not make much sense.

File: radix-engine-tests/assets/blueprints/generic_proxy/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+            let oracle_address = self.component_address.unwrap();
+            let args = scrypto_encode(&args).unwrap();
+
+            let bytes = ScryptoVmV1Api::object_call(
+                oracle_address.address().as_node_id(),
+                &method_name,
+                args,
+            );
+            scrypto_decode(&bytes).unwrap()
+        }
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/generic_proxy/src/lib.rs]: Good point, thanks :)

File: radix-engine-tests/assets/blueprints/generic_proxy/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        //         "proxy_call",
+        //         manifest_args!(
+        //             "get_price",
+        //             to_manifest_value(&("XRD".to_string(),)).unwrap()
+        //         ),
+        //     )
+        //     .build();
+        //  ```
+        pub fn proxy_call(&self, method_name: String, args: ScryptoValue) -> ScryptoValue {
+            let oracle_address = self.component_address.unwrap();
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/generic_proxy/src/lib.rs]: Done

File: radix-engine-tests/assets/blueprints/Cargo.toml [hunk 8]
```diff
@@ -66,7 +66,12 @@ members = [
     "system_wasm_buffers",
     "event-replacement",
     "decimal",
-    "crypto_scrypto"
+    "crypto_scrypto",
+    "oracle_proxy_basic",
+    "generic_proxy",
```
[the remark this answers is not in the document]
Comment (lrubasze) [reply radix-engine-tests/assets/blueprints/Cargo.toml]: Done

**Pull request merged**